File: server/agent/src/utils/places-photos.js
/**
 * Google Places Photos — Utility for fetching real place photos
 *
 * Uses the Google Places API (New) to:
 * 1. Search for a place by name + city → get Place ID
 * 2. Get photo references for that place
 * 3. Build proxied photo URLs for the frontend
 *
 * Uses Node.js native fetch (available since Node 18+).
 */

require('dotenv').config();

const API_KEY = process.env.GOOGLE_MAPS_API_KEY;
const PLACES_BASE = 'https://places.googleapis.com/v1';

/**
 * Search for a place using Google Places Text Search (New).
 * Returns the first result's Place ID and photo references.
 *
 * @param {string} placeName — e.g. "Krishna Janmabhoomi"
 * @param {string} city — e.g. "Mathura"
 * @returns {Promise<{ placeId: string|null, photoRef: string|null, displayName: string|null }>}
 */
async function searchPlace(placeName, city) {
    if (!API_KEY) {
        return { placeId: null, photoRef: null, displayName: null };
    }

    try {
        const query = `${placeName}, ${city}, India`;

        const res = await fetch(`${PLACES_BASE}/places:searchText`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-Goog-Api-Key': API_KEY,
                'X-Goog-FieldMask': 'places.id,places.displayName,places.photos',
            },
            body: JSON.stringify({
                textQuery: query,
                languageCode: 'en',
                maxResultCount: 1,
            }),
        });

        if (!res.ok) {
            console.warn(`[PlacesPhotos] Search failed for "${query}": ${res.status}`);
            return { placeId: null, photoRef: null, displayName: null };
        }

        const data = await res.json();
        const place = data.places?.[0];

        if (!place) {
            return { placeId: null, photoRef: null, displayName: null };
        }

        const photoRef = place.photos?.[0]?.name || null;
        return {
            placeId: place.id || null,
            photoRef,
            displayName: place.displayName?.text || null,
        };
    } catch (err) {
        console.error(`[PlacesPhotos] Error searching "${placeName}":`, err.message);
        return { placeId: null, photoRef: null, displayName: null };
    }
}

/**
 * Fetch a place photo from Google Places (New) API.
 * Returns the raw image response (stream) for proxying.
 *
 * @param {string} photoReference — The photo resource name (e.g. "places/xxx/photos/yyy")
 * @param {number} maxWidth — Max width in pixels (default 400)
 * @returns {Promise<{ ok: boolean, stream: ReadableStream|null, contentType: string|null }>}
 */
async function fetchPhoto(photoReference, maxWidth = 400) {
    if (!API_KEY || !photoReference) {
        return { ok: false, stream: null, contentType: null };
    }

    try {
        const url = `${PLACES_BASE}/${photoReference}/media?maxWidthPx=${maxWidth}&key=${API_KEY}`;

        const res = await fetch(url, { redirect: 'follow' });

        if (!res.ok) {
            console.warn(`[PlacesPhotos] Photo fetch failed: ${res.status}`);
            return { ok: false, stream: null, contentType: null };
        }

        return {
            ok: true,
            stream: res.body,
            contentType: res.headers.get('content-type') || 'image/jpeg',
        };
    } catch (err) {
        console.error(`[PlacesPhotos] Photo fetch error:`, err.message);
        return { ok: false, stream: null, contentType: null };
    }
}

/**
 * Build the proxied photo URL for the frontend.
 *
 * @param {string} placeId — The internal place ID (e.g. "mathura_001")
 * @param {number} maxWidth — optional max width
 * @returns {string}
 */
function buildPhotoUrl(placeId, maxWidth = 400) {
    return `/api/agent/places/${placeId}/photo?maxwidth=${maxWidth}`;
}

module.exports = { searchPlace, fetchPhoto, buildPhotoUrl };

File: server_py/agent/utils/places_photos.py
"""
Google Places Photos — Utility for fetching real place photos.
"""

import httpx
from config import get_settings

PLACES_BASE = "https://places.googleapis.com/v1"


async def search_place(place_name: str, city: str) -> dict:
    """Search for a place using Google Places Text Search (New).

    Returns:
        dict with placeId, photoRef, displayName (or None values)
    """
    settings = get_settings()
    api_key = settings.GOOGLE_MAPS_API_KEY
    if not api_key:
        return {"placeId": None, "photoRef": None, "displayName": None}

    try:
        query = f"{place_name}, {city}, India"
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{PLACES_BASE}/places:searchText",
                headers={
                    "Content-Type": "application/json",
                    "X-Goog-Api-Key": api_key,
                    "X-Goog-FieldMask": "places.id,places.displayName,places.photos",
                },
                json={
                    "textQuery": query,
                    "languageCode": "en",
                    "maxResultCount": 1,
                },
            )

        if resp.status_code != 200:
            print(f'[PlacesPhotos] Search failed for "{query}": {resp.status_code}')
            return {"placeId": None, "photoRef": None, "displayName": None}

        data = resp.json()
        place = (data.get("places") or [None])[0]
        if not place:
            return {"placeId": None, "photoRef": None, "displayName": None}

        photos = place.get("photos") or []
        photo_ref = photos[0].get("name") if photos else None

        return {
            "placeId": place.get("id"),
            "photoRef": photo_ref,
            "displayName": (place.get("displayName") or {}).get("text"),
        }
    except Exception as e:
        print(f'[PlacesPhotos] Error searching "{place_name}": {e}')
        return {"placeId": None, "photoRef": None, "displayName": None}


async def fetch_photo(photo_reference: str, max_width: int = 400) -> dict:
    """Fetch a place photo from Google Places (New) API.

    Returns:
        dict with ok, content (bytes), content_type
    """
    settings = get_settings()
    api_key = settings.GOOGLE_MAPS_API_KEY
    if not api_key or not photo_reference:
        return {"ok": False, "content": None, "content_type": None}

    try:
        url = f"{PLACES_BASE}/{photo_reference}/media?maxWidthPx={max_width}&key={api_key}"
        async with httpx.AsyncClient(follow_redirects=True) as client:
            resp = await client.get(url)

        if resp.status_code != 200:
            print(f"[PlacesPhotos] Photo fetch failed: {resp.status_code}")
            return {"ok": False, "content": None, "content_type": None}

        return {
            "ok": True,
            "content": resp.content,
            "content_type": resp.headers.get("content-type", "image/jpeg"),
        }
    except Exception as e:
        print(f"[PlacesPhotos] Photo fetch error: {e}")
        return {"ok": False, "content": None, "content_type": None}


def build_photo_url(place_id: str, max_width: int = 400) -> str:
    """Build the proxied photo URL for the frontend."""
    return f"/api/agent/places/{place_id}/photo?maxwidth={max_width}"
